test/hotspot/gtest/concurrentTestRunner.inline.hpp

```diff
@@ -0,0 +1,100 @@
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+#ifndef GTEST_CONCURRENT_TEST_RUNNER_INLINE_HPP
+#define GTEST_CONCURRENT_TEST_RUNNER_INLINE_HPP
+
+#include "threadHelper.inline.hpp"
+
+// This file contains helper classes to run unit tests concurrently in multiple threads.
+
+// Base class for test runnable. Override runUnitTest() to specify what to run.
+class TestRunnable {
+public:
+  virtual void runUnitTest() {
+  };
+};
+
+// This class represents a thread for a unit test.
+class UnitTestThread : public JavaTestThread {
+public:
+  // runnableArg - what to run
+  // doneArg - a semaphore to notify when the thread is done running
+  // testDurationArg - how long to run (in milliseconds)
+  UnitTestThread(TestRunnable* runnableArg, Semaphore* doneArg, long testDurationArg) : JavaTestThread(doneArg) {
+    runnable = runnableArg;
+    testDuration = testDurationArg;
+  }
+
+  virtual ~UnitTestThread() {}
+
+  // from JavaTestThread
+  void main_run() {
+    long stopTime = os::javaTimeMillis() + testDuration;
+    while (os::javaTimeMillis() < stopTime) {
+      runnable->runUnitTest();
+    }
+  }
+private:
+  long testDuration;
+  TestRunnable* runnable;
+};
+
+// Helper class for running a given unit test concurrently in multiple threads.
+class ConcurrentTestRunner {
+public:
+  // runnableArg - what to run
+  // nrOfThreadsArg - how many threads to use concurrently
+  // testDurationMillisArg - duration for each test run
+  ConcurrentTestRunner(TestRunnable* runnableArg, int nrOfThreadsArg, long testDurationMillisArg) {
+    unitTestRunnable = runnableArg;
+    nrOfThreads = nrOfThreadsArg;
+    testDurationMillis = testDurationMillisArg;
+  }
+
+  virtual ~ConcurrentTestRunner() {}
+
+  void run() {
+    Semaphore done(0);
+
+    std::vector<UnitTestThread*> t;
+    for (int i = 0; i < nrOfThreads; i++) {
+      t.push_back(new UnitTestThread(unitTestRunnable, &done, testDurationMillis));
+    }
+
+    for (int i = 0; i < nrOfThreads; i++) {
+      t[i]->doit();
+    }
+
+    for (int i = 0; i < nrOfThreads; i++) {
+      done.wait();
+    }
+  }
+
+private:
+  long testDurationMillis;
+  int nrOfThreads;
+  TestRunnable* unitTestRunnable;
+};
+
+#endif // include guard
```

Comment:
we tend to use the expression used in the corresponding `#if` / `#ifdef` as a comment in `#endif`

Reply:
"include guard" is self evident. We usually write the name of it, like this:
```
#endif // GTEST_CONCURRENT_TEST_RUNNER_INLINE_HPP
```